### DISCUSSION TITLE
Unit testing business logic in maui project

### DISCUSSION THREAD

**Discussion by fgma** [dotnet/maui#5828] [Q&A]
I'm just getting started with maui and try to test my business logic. When adding a test project into my VS solution I can't reference the maui project for testing. I've also tried moving my business logic to a separate class library and test it there which also doesn't work. I'm always getting errors about unsupported target frameworks "net6.0-windows10.0.19041".

I don't care which framework to use at the moment.  Just trying to get some tests running. I'm using latest VS2022 preview and started from the default code for a new project adding just a single class to it.

What is the way to go for unit testing with maui?

**Comment by jamesmontemagno** [comment]
One option is to break your logic you want to test out into another class library that targets just net6.0 and then you add that as a reference to your .NET MAUI app and your Unit Test project

**Reply by southsideringo** [reply]
This would be an excellent subject for a future video on your YouTube channel.

**Reply by kkohler2** [reply]
This won't really work.  I have Converters that I have Xamarin.Forms tests for that work fine, but can't get working in Maui.

So something like:

        [Test]
        public void Convert()
        {
            // Arrange
            string rgbColor = "#ff0000";

            // Act
            var result = _rgbToColorConverter.Convert(rgbColor, null, null, null);

            // Assert
            Assert.AreEqual(Colors.Red, result);
        }

Except that Colors.Red requires use of Microsoft.Maui.Graphics, thus the unit test project needs to be a Maui project, but targeting net6.0.

I've tried updating my actual project by:
Adding target of **net6.0**
And updating the project file with the following, so it will compile as a library and an application:
  **&lt;PropertyGroup Condition=&quot;&#39;$(TargetFramework)&#39; == &#39;net6.0&#39;&quot;&gt;
    &lt;OutputType&gt;Library&lt;/OutputType&gt;
  &lt;/PropertyGroup&gt;**


So the scenario is:
1. Maui project w/ net6.0 target added and property group for net6.0 to compile as a library.
2. Maui unit test project (maui class library project w/ Microsoft.NET.Test.Sdk, NUnit, and NUnit3TestAdapter nuget packages, plus the unit tests)  This project references the Maui project that creates the app.

Build the test project and try to run unit tests.  I get errors for the results:

Severity	Code	Description	Project	File	Line	Suppression State
Error		Empty ResolveFrameworkReference.RuntimePackPath while trying to read runtime components manifest. ResolvedFrameworkReference available: { Microsoft.NETCore.App, RuntimePackPath:  }	MauiTestProject	C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.MonoTargets.Sdk\6.0.9\Sdk\RuntimeComponentManifest.targets	3	

Severity	Code	Description	Project	File	Line	Suppression State
Error		Empty ResolveFrameworkReference.RuntimePackPath while trying to read runtime components manifest. ResolvedFrameworkReference available: { Microsoft.NETCore.App, RuntimePackPath:  }	MauiTestProject	C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.MonoTargets.Sdk\6.0.9\Sdk\RuntimeComponentManifest.targets	3	

Anyone know how to fix these errors?

**Comment by fgma** [comment]
@jamesmontemagno I needed to create a new Maui library project and manually added an additional net6.0 target by editing the project. A normal .net core class library doesn't work because I can't reference it in the maui project anymore.

**Comment by Fabbrik1976** [comment]
I added  `<UseMaui>true</UseMaui>` to my xunit project and then I was able to add the project to my MAUI solution. 

![image](https://user-images.githubusercontent.com/86317458/170942461-cd53d9d9-2704-4611-b164-5a026973ba42.png)

**Comment by Craige** [comment]
I'm also having a similar problem with unit testing a Maui project with XUnit. I've created the test project, added net6.0 as a target framework to my main project, added `<UseMaui>` to the test project, and added a reference to the Maui project to the test project, but upon running `dotnet test` I get the following:

`CSC : error CS5001: Program does not contain a static 'Main' method suitable for an entry point [.../path/to/MyMauiProject.csproj]`

Creating a `Main` entry point in my Maui project does indeed allow the tests to execute, but breaks the Maui project build because the compiler gives the Main entry point priority over the `MauiProgram.CreateMauiApp` entry point.

Right now I can test by leaving the entry `Main` entry point in my code base and commenting or un-commenting it for testing/build, but that's obviously not a solution. What am I missing here?

**Reply by NeilMalcolm** [reply]
Try adding the following PropertyGroup to your csproj:

```xml
    <PropertyGroup Condition="'$(TargetFramework)' == 'net6.0'">
        <OutputType>Library</OutputType>
    </PropertyGroup>
```

**Reply by Craige** [reply]
> Try adding the following PropertyGroup to your csproj [...]

That seems to be the element that I was missing. I need to output as a library when unit testing.

Thank you very much!

**Comment by TrueGeek** [comment]
I've got this working with the same trick that Fabbrik1976 showed. However, when I run

`dotnet test --collect: "XPlat Code Coverage"`

it shows 0 lines / branches in coverage.cobertura.xml. Anyone else having this problem?

**Reply by RichardMarsh** [reply]
> I've got this working with the same trick that Fabbrik1976 showed. However, when I run
> 
> `dotnet test --collect: "XPlat Code Coverage"`
> 
> it shows 0 lines / branches in coverage.cobertura.xml. Anyone else having this problem?

Yup, same issue here.

**Comment by MahadevBansode** [comment]
Did anyone is able to fix these code coverage issue ? I am also facing same issue
